lgc/elfLinker/ElfLinker.cpp

```diff
@@ -0,0 +1,645 @@
+// @param inputSectIdx : Index of input section within this output section that the symbol refers to
+void OutputSection::addSymbol(const object::ELFSymbolRef &elfSymRef, unsigned inputSectIdx) {
+  const InputSection &inputSection = m_inputSections[inputSectIdx];
+  StringRef name = cantFail(elfSymRef.getName());
+  ELF::Elf64_Sym newSym = {};
+  newSym.st_name = m_linker->getStringIndex(name);
+  newSym.setBinding(elfSymRef.getBinding());
+  newSym.setType(cantFail(elfSymRef.getType()));
+  newSym.st_shndx = getIndex();
+  newSym.st_value = cantFail(elfSymRef.getValue()) + inputSection.offset;
+  newSym.st_size = elfSymRef.getSize();
+  if (m_linker->findSymbol(newSym.st_name) != 0)
+    report_fatal_error("Duplicate symbol '" + name + "'");
+  m_linker->getSymbols().push_back(newSym);
+}
+
+// =====================================================================================================================
+// Write the output section
+//
+// @param [in/out] outStream : Stream to write the section to
+// @param [in/out] shdr : ELF section header to write to (but not sh_offset)
+void OutputSection::write(raw_pwrite_stream &outStream, ELF::Elf64_Shdr *shdr) {
+  shdr->sh_name = m_linker->getStringIndex(getName());
+  m_offset = outStream.tell();
+
+  if (m_type == ELF::SHT_STRTAB) {
+    StringRef strings = m_linker->getStrings();
+    shdr->sh_type = m_type;
+    shdr->sh_size = strings.size();
+    m_linker->setStringTableIndex(getIndex());
+    outStream << strings;
+    return;
+  }
+
+  if (m_type == ELF::SHT_SYMTAB) {
+    ArrayRef<ELF::Elf64_Sym> symbols = m_linker->getSymbols();
+    shdr->sh_type = m_type;
+    shdr->sh_size = symbols.size() * sizeof(ELF::Elf64_Sym);
+    shdr->sh_entsize = sizeof(ELF::Elf64_Sym);
+    shdr->sh_link = 1; // Section index of string table
+    outStream << StringRef(reinterpret_cast<const char *>(symbols.data()), symbols.size() * sizeof(ELF::Elf64_Sym));
+    return;
+  }
+
+  if (m_type == ELF::SHT_NOTE) {
+    StringRef notes = m_linker->getNotes();
+    shdr->sh_type = m_type;
+    shdr->sh_size = notes.size();
+    outStream << notes;
+    return;
+  }
+
+  if (m_inputSections.empty())
+    return;
+
+  // This section has contributions from input sections. Get the type and flags from the first input section.
+  shdr->sh_type = object::ELFSectionRef(m_inputSections[0].sectionRef).getType();
+  shdr->sh_flags = object::ELFSectionRef(m_inputSections[0].sectionRef).getFlags();
+
+  // Set up the pattern we will use for alignment padding.
+  const size_t paddingUnit = 16;
+  const char *padding = "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0";
+  bool paddingAtEnd = false;
+  if (shdr->sh_flags & ELF::SHF_EXECINSTR) {
+    if (m_linker->getPipelineState()->getTargetInfo().getGfxIpVersion().major < 10)
+      padding = "\0\0\x80\xBF\0\0\x80\xBF\0\0\x80\xBF\0\0\x80\xBF"; // s_nop
+    else {
+      padding = "\0\0\x9F\xBF\0\0\x9F\xBF\0\0\x9F\xBF\0\0\x9F\xBF"; // s_code_end
+      paddingAtEnd = true;
+    }
+  }
```

Comment:
Presumably this needs adjustment when we start having glue code? There's still reason to align to a higher multiple (e.g. for loop alignment), but obviously it needs an s_nop padding between glue parts on all hardware...

Reply:
Yes, I think you're right. Glue code also needs additional handling: it needs to remove the final s_endpgm in some way (which also requires constant pools to be in a different rodata section, not done yet).

Reply:
We should just use the existing "return" facility for that: for Mesa support, LLVM already knows not to emit `s_endpgm` for shaders that return non-void, and it knows to put the return basic block last in ISA order. (The SI_RETURN_TO_EPILOG pseudo-instruction is used in that case.)

Reply:
Yeah, I spotted that whilst working on the fetch shader, and I assumed it was a deliberate policy for Mesa in the LLVM back-end (without actually checking).

But it still seems to add the s_code_end padding for GFX10, which is a shame. Do you think we should fix the back-end to not do that?

Reply:
Yes, fixing the backend for that would be good. I think it's already fixed for Mesa, but contingent on the OS in the triple. For the amdpal OS, the question boils down to whether we want to build a relocatable ELF or not... I haven't thought about how to best express that. Of course, we could just heuristically scan if there are any non-void-returning functions in the module and decide based on that...

Reply:
The back-end fix is D80338, in review.